File: Lua/trie.lua

function getTrieNode() return {children={}, isEnd=false} end

Trie = {}  Trie.__index = Trie

function Trie:new()
  local tr = {}
  setmetatable(tr, Trie)
  tr.root = getTrieNode()
  tr.suggestions = {}
  return tr
end

function Trie:form(...)
  for i, v in ipairs(arg[1]) do
    self:insert(v)
  end
end

function Trie:insert(key)
  node = self.root

  for i=1, #key do
    local c = key:sub(i,i)


    if node.children[c] == nil then
      node.children[c] = getTrieNode()
    end

    node = node.children[c]
    
  end

  node.isEnd = true
  
end

function Trie:suggestionsRec(node, word)
  if node.isEnd then
    self.suggestions[#self.suggestions+1] = word
  end

  for a, n in pairs(node.children) do
    self:suggestionsRec(n, word..a)
  end
end

function Trie:getAutoSugg(key)
  
  self.suggestions = {}
  node = self.root
  
  for i=1, #key do
    local c = key:sub(i,i)
    if node.children[c] == nil then return 0 end
    node = node.children[c]
  end

  if node.children == nil then return -1 end

  self:suggestionsRec(node, key)

  return 1

end

File: Python/trie.py
#! /usr/bin/env python3.10

#NOTE may be able to represent TrieNode as a dict {..., isEnd: 0}

class TrieNode:
	def __init__(self):
		self.children = {}
		self.isEnd = False

#NOTE try to add weights to letters on a branch 
# weight n = len(b.children)

class Trie:
	def __init__(self):
		self.root = TrieNode()
		self.suggestions = []
	
	def formTrie(self, keys):
		for key in keys:
			self.insert(key)
	
	def insert(self, key):
		node = self.root

		for c in key:
			if not node.children.get(c):
				node.children[c] = TrieNode()

			node = node.children[c]

		node.isEnd = True
	
	def suggestionRec(self, node, word):
		if node.isEnd:
			self.suggestions.append(word)

		for a, n in node.children.items():
			self.suggestionRec(n, word + a)
	
	def getAutoSugg(self, key):
		self.suggestions = []
		node = self.root
		for c in key:
			if not node.children.get(c):
				return 0
			node = node.children[c]

		if not node.children:
			return -1

		self.suggestionRec(node, key)

		return 1

